[DOCS] Overhaul TLS security docs (#68946)
elastic/elasticsearch: x-pack/docs/en/security/securing-communications/security-basic-setup-https.asciidoc

addition layer of security ensures that all communications to and from your
cluster are secured.

When you run the `elasticsearch-certutil` tool in `http` mode, the tool asks
several questions about how you want to generate certificates. While there are
numerous options, the following choices result in certificates that should
work for most environments.

[[signing-certificates]]
.Signing certificates
****
The first question that the `elasticsearch-certutil` tool prompts you with is
whether you want to generate a Certificate Signing Request (CSR). Answer
`n` if you want to sign your own certificates, or `y` if you want to sign
certificates with a central CA.

[discrete]
===== Sign your own certificates

If you want to use the CA that you created when
<<generate-certificates,Generating the certificate authority>> answer `n` to
the first question. You then specify the location of your CA, which the tool
uses to sign and generate a `.p12` certificate. The steps in this procedure
follow this workflow.

[discrete]
===== Sign certificates with a central CA

If you work in an environment with a central security team, they can likely

> generate a certificate for you. Infrastructure within your organization

Reviewer: Nit: I think there is a difference between "generate" and "sign" a certificate. If the central security team can just generate the certificate, we don't even need to use the `elasticsearch-certutil` tool at all. The tool is relevant if the security team "signs" the certificate for which we need the CSR file.

might already be configured to trust an existing CA, so it may be easier
for clients to connect to {es} if you use a CSR and send that
request to the team that controls your CA. To use a central CA, answer `y` to
the first question.
****

[[basic-setup-https-prerequisites]]
==== Prerequisites

Complete all steps in <<security-basic-setup,Set up basic security for the Elastic Stack>>.

[[encrypt-http-communication]]
==== Encrypt HTTP client communications for {es}

. Stop {es} and {kib} if they are running.

. From the directory where you installed {es}, run the {es}

>    HTTP certificate tool to generate a Certificate Signing Request (CSR).

Reviewer: The following steps are actually for generating the certificates, not the CSR.

+
[source,shell]
----
./bin/elasticsearch-certutil http
----
+
This command generates a `.zip` file that contains certificates and keys
to use with {es} and {kib}. Each folder contains a `README.txt`
explaining how to use these files.

   a. When asked if you want to generate a CSR, enter `n`.

   b. When asked if you want to use an existing CA, enter `y`.

   c. Enter the path to your CA. This is the absolute path to

>    the `elastic-stack-ca.p12` file that you generated for your cluster.

Reviewer: This can work. But I think we recommend to not share the CA between the transport and the http layers. The CA for transport layer should not be used for anything else since TLS is the only protection the inter-node communication has.

   d. Enter the password for your CA.

   e. Enter an expiration value for your certificate. You can enter the
   validity period in years, months, or days. For example, enter `90D` for 90
   days.

   f. When asked if you want to generate one certificate per node, enter `y`.
+
Each certificate will have its own private key, and will be issued for a
specific hostname or IP address.

>    g. When prompted, enter the name of the first node in your cluster. Use the same node name from previous steps.

Author: Good catch! This is the same node name used when [generating node certificates](https://elasticsearch_68946.docs-preview.app.elstc.co/guide/en/elasticsearch/reference/master/security-basic-setup.html#generate-certificates). I'll add a clarification and a link to that section.

   h. Enter all hostnames used to connect to your first node. These hostnames
   will be added as DNS names in the Subject Alternative Name (SAN) field in your certificate.
+
List every hostname and variant used to connect to your cluster over HTTPS.

   i. Enter the IP addresses that clients can use to connect to your node.

   j. Repeat these steps for each additional node in your cluster.

. After generating a certificate for each of your nodes, enter a password for
   your private key when prompted.

. Unzip the generated `elasticsearch-ssl-http.zip` file. This compressed file
   contains one directory for both {es} and {kib}.
+
--
[source,txt]
----
/elasticsearch
|_ README.txt
|_ http.p12

> |_ sample-elasticsearch.yml

Reviewer: This directory layout is when we enter `n` for the question of "do you want to generate one certificate per node". But the previous step `f` says enter `y` to this question.

----

[source,txt]
----
/kibana
|_ README.txt
|_ elasticsearch-ca.pem
|_ sample-kibana.yml
----
--

. Copy the relevant `http.p12` certificate to the `ES_PATH_CONF` directory on each node.

. On each node, edit the `elasticsearch.yml` file to enable HTTPS security and
   specify the location of the `http.p12` security certificate.
+
[source,yaml]
----
xpack.security.http.ssl.enabled: true
xpack.security.http.ssl.keystore.path: http.p12
----

> . Add the password for your private key to a secure setting in {es}.

Author: I like "to the secure settings" best.

+
[source,shell]
----
./bin/elasticsearch-keystore add xpack.security.http.ssl.keystore.secure_password
----

> . Restart {es}.

Author: Agreed! I will change this to:
>Start Elasticsearch.

**Next**: <<encrypt-kibana-http,Encrypt HTTP client communications for {kib}>>

[[encrypt-kibana-http]]
==== Encrypt HTTP client communications for {kib}

Browsers send traffic to {kib} and {kib} sends traffic to {es}.
These communication channels are configured separately to use TLS. You encrypt
traffic between your browser and {kib}, and then encrypt traffic between
{kib} and {es}.

[[encrypt-kibana-browser]]
===== Encrypt traffic between your browser and {kib}

You create a server certificate and private key for {kib}. {kib} uses this
server certificate and corresponding private key when receiving connections
from web browsers.

When you obtain a server certificate, you must set its subject alternative

> name (SAN) correctly to ensure that browsers with hostname verification will

Author: I'll remove the "with hostname verification" part so that the sentence reads:
>When you obtain a server certificate, you must set its subject alternative name (SAN) correctly to ensure that browsers will trust it.

trust it. You can set one or more SANs to the {kib} server’s fully-qualified
domain name (FQDN), hostname, or IP address. When choosing the SAN, pick
whichever attribute you'll use to connect to {kib} in your browser, which is
likely the FQDN.

> The following instructions create an untrusted browser certificate for {kib}.

Reviewer: This bit is confusing to me: it says "an untrusted browser certificate" but the following texts suggest we send the CSR file to the internal CA or trusted CA for signing, which means the certificate will be trusted. I am not sure what we are trying to say here?

Author: Yeah, that language is a little soft. @ywangd, is this more clear?

> The following instructions create a Certificate Signing Request (CSR) for {kib}. A CSR contains information that a CA uses to generate and sign a security certificate. A CSR is acceptable for development environments and building a proof of concept, but should not be used in a production environment.

Reviewer: CSR is not a certificate. A certificate will be generated once a CSR is signed. Depending who signs the CSR, the certificate can be either a trusted certificate (because a public trusted CA signs it) or an untrusted certificate (because an internal CA signs it). I think what you meant is that an self generated or internally signed certificate is acceptable for development, but less ideal for production and not acceptable if the production environment is publicly accessible.

Author: Thank you for the clarification @ywangd! I think that this revision captures the differences between trusted/untrusted and how a CSR is used to generate a certificate:

> The following instructions create a Certificate Signing Request (CSR) for {kib}. A CSR contains information that a CA
> uses to generate and sign a security certificate. The certificate can be trusted (signed by a public, trusted CA)
> or untrusted (signed by an internal CA). A self-signed or internally-signed certificate is acceptable for development
> environments and building a proof of concept, but should not be used in a production environment.

This certificate is acceptable for development environments and building a
proof of concept, but should not be used in a production environment.

WARNING: Before going to production, use a trusted CA such as https://letsencrypt.org/[Let's
Encrypt] or your organization's internal CA to sign the certificate. Using a
signed certificate establishes browser trust for connections to {kib} for
internal access or on the public internet.

. Generate a server certificate and private key for {kib}.
+
[source,shell]
----
./bin/elasticsearch-certutil csr -name kibana-server -dns example.com,www.example.com
----
+
The CSR has a common name (CN) of `kibana-server`, a SAN of `example.com`,
and another SAN of `www.example.com`.
+
This command generates a `csr-bundle.zip` file by default with the following
contents:
+
[source,txt]
----
/kibana-server
|_ kibana-server.csr
|_ kibana-server.key
----

. Unzip the `csr-bundle.zip` file to obtain the `kibana-server.csr` unsigned
security certificate and the `kibana-server.key` unencrypted private key.

> . Send the `kibana-server.csr` unsigned security certificate to your internal

Reviewer 2: ```suggestion
. Send the `kibana-server.csr` certificate signing request to your internal
```

Author: I'll change this to "certificate signing request".

> CA or trusted CA for signing. The signed file should be a `.crt` file, such as

Reviewer 2: The suffix doesn't matter here. It can be `.crt`, `.pem`, `.cer` ( or `.yangandioannishavetoomanynitcomments` :) ). In addition to Yang's comment below, I think that we can remove this sentence and the following

Author: I removed the `IMPORTANT` admonition to instead focus on this text. I think we can revise and at least inform users that the file can be in different types, and then provide an example of what type it might be:
>. Send the `kibana-server.csr` certificate signing request to your internal CA or trusted CA for signing to obtain a signed certificate. The signed file can be in different formats, such as a `.crt` file like `kibana-server.crt`.

`kibana-server.crt`.
+

> WARNING: {kib} will not trust the CSR. You must send the `kibana-server.csr`

Reviewer: CSR is not a certificate. So technically Kibana just cannot use it as one, it is not about trust or not.

Author: Thanks for the clarification! I can change to read as:

> Kibana cannot use the CSR. You must send the `kibana-server.csr` unsigned security certificate to your internal CA or trusted CA for signing to obtained a signed certificate.

> unsigned security certificate to your internal CA or trusted CA for

Reviewer 2: It's a certificate signing request , not an unsigned certificate

signing.

. Open `kibana.yml` and add the following lines to configure {kib} to access
the server certificate and unencrypted private key.
+
[source,yaml]
----
server.ssl.certificate: KBN_PATH_CONF/kibana-server.crt
server.ssl.key: KBN_PATH_CONF/kibana-server.key
----
+
NOTE: `KBN_PATH_CONF` contains the path for the {kib} configuration files. If
you installed {kib} using archive distributions (`zip` or `tar.gz`), the
path defaults to `KBN_HOME/config`. If you used package distributions
(Debian or RPM), the path defaults to `/etc/kibana`.

. Add the following line to `kibana.yml` to enable TLS for inbound
connections.
+
[source,yaml]
----
server.ssl.enabled: true
----

> . Restart {kib}.

Reviewer 2: Again, maybe just "Start" ?

Author: Yes -- I will change to "Start Kibana."

NOTE: After making these changes, you must always access {kib} via HTTPS. For
example, `https://<your_kibana_host>.com`.

**Next**: <<encrypt-kibana-elasticsearch,Encrypt traffic between {kib} and {es}>>

[[encrypt-kibana-elasticsearch]]

> ===== Encrypt traffic between {kib} and {es}

Reviewer: I think this sub-section should be placed before the previous one. TLS between browser and Kibana is arguably more challenging compared to this one.

Author: ++ I'll swap those sections.

When you ran the `elasticsearch-certutil` tool with the `http` option, it
created a `/kibana` directory containing an `elasticsearch-ca.pem` file. You
use this file to configure {kib} to trust the {es} CA for the HTTP
layer.

1. Copy the `elasticsearch-ca.pem` file to the {kib} configuration directory,
as defined by the `KBN_PATH_CONF` path.

2. Open `kibana.yml` and add the following line to specify the location of the
security certificate for the HTTP layer.
+
[source,yaml]
----
elasticsearch.ssl.certificateAuthorities: KBN_PATH_CONF/elasticsearch-ca.pem
----

3. Add the following line to specify the HTTPS URL for your {es}
cluster.
+
[source,yaml]
----
elasticsearch.hosts: https://<your_elasticsearch_host>.com:9200
----

4. Restart {kib}.

.Connect to a secure monitoring cluster
****
If the Elastic monitoring features are enabled and you configured a separate

> {es} monitoring cluster, you can also configure {kib} to connect to

Reviewer 2: Maybe worth mentioning that the monitoring cluster will have it's own elasticsearch-ca.pem file?

Author: I'll add the following note:
>NOTE: You must create a separate `elasticsearch-ca.pem` security file for the monitoring cluster.

[discrete]
[[beats-monitoring-role]]
====== Create a monitoring role

To send monitoring data securely, create a monitoring user and grant it the
necessary privileges.

You can use the built-in `beats_system` user, if it’s available in your
environment. Because the built-in users are not available in Elastic Cloud,
these instructions create a user that is explicitly used for monitoring
{metricbeat}.

1. Create the monitoring role:

   a. Enter **metricbeat_monitoring** as the role name.

   b. Choose the **monitor** cluster privilege.

   c. On the **.monitoring-beats-\*** indices, choose the **create_index** and
   **create_doc** privileges.

2. Create the monitoring user:

   a. Enter **metricbeat_monitoring** as the user name.

   b. Enter the username, password, and other user details.

   c. Assign the following roles to the **metricbeat_monitoring** user:
+
[cols="1,1"]
|===
| Role                    | Purpose

| `metricbeat_monitoring` | Monitor {metricbeat}.
| `kibana_admin`          | Use {kib}
| `monitoring_user`       | Use Stack Monitoring in {kib} to monitor {metricbeat}
|===

**Next**: <<beats-writer-role,Create a writer role>>

[discrete]
[[beats-writer-role]]

> ====== Create a writer role

Reviewer 2: ```suggestion
====== Create a writer role and user
```

Users who publish events to {es} need to create and write to {metricbeat} indices. To minimize the privileges required by the writer role, use the setup role to pre-load dependencies. This section assumes that you’ve
<<beats-setup-role,created the setup role>>.

1. Create the writer role:

   a. Enter **metricbeat_writer** as the role name.

   b. Choose the **monitor** and **read_ilm** cluster privileges.

   c. On the **metricbeat-\*** indices, choose the **create_doc**, **create_index**, and **view_index_metadata** privileges.

2. Create the writer user:

   a. Enter **metricbeat_writer** as the user name.

   b. Enter the username, password, and other user details.

   c. Assign the following roles to the **metricbeat_writer** user:
+
[cols="1,1"]
|===
| Role                          | Purpose

| `metricbeat_writer`           | Monitor {metricbeat}
| `remote_monitoring_collector` | Collect monitoring metrics from {metricbeat}
| `remote_monitoring_agent`     | Send monitoring data to the monitoring cluster
|===

**Next**: <<beats-reader-role,Create a reader role>>

[discrete]
[[beats-reader-role]]

> ====== Create a reader role

Reviewer 2: ```suggestion
====== Create a reader role and user
```

   - Choose **Read** or **All** for Discover, Visualize, Dashboard, and Metrics.

2. Create the reader user:

   a. Enter **metricbeat_reader** as the user name.

   b. Enter the username, password, and other user details.

   c. Assign the following roles to the **metricbeat_reader** user:
+
[cols="1,1"]
|===
| Role                          | Purpose

| `metricbeat_reader` | Read {metricbeat} data.
| `monitoring_user`   | Allow users to monitor the health of {metricbeat}
itself. Only assign this role to users who manage {metricbeat}
| `beats_admin`       | Create and manage configurations in {beats} central
management. Only assign this role to users who need to use {beats} central
management.
|===

**Next**: <<configure-metricbeat-tls,Configure {metricbeat} to use TLS>>

[discrete]
[[configure-metricbeat-tls]]
===== Configure {metricbeat} to use TLS

Before starting {metricbeat}, you configure the connections to {es} and
Kibana. You can configure authentication to send data to your secured cluster
using basic authentication, API key authentication, or Public Key
Infrastructure (PKI) certificates.

The following instructions use the credentials for the `metricbeat_writer`

> and `metricbeat_setup` roles that you created. If you need a greater level of

Reviewer 2: ```suggestion
and `metricbeat_setup` users that you created. If you need a greater level of
```

Author: I'll change "roles" to "users".

After configuring connections to Elasticsearch and Kibana, you'll enable the
`elasticsearch-xpack` module and configure that module to use HTTPS.

WARNING: In production environments, we strongly recommend using a separate
cluster (referred to as the monitoring cluster) to store your data. Using a
separate monitoring cluster prevents production cluster outages from impacting
your ability to access your monitoring data. It also prevents monitoring
activities from impacting the performance of your production cluster.

. From the directory where you installed Elasticsearch, navigate to the
`/kibana` directory that you created when <<encrypt-http-communication,encrypting HTTP client communications for {es}>>.

. Copy the `elasticsearch-ca.pem` certificate to the directory where you
installed Metricbeat.

. Open the `metricbeat.yml` configuration file and configure the connection
to Elasticsearch.
+
Under `output.elasticsearch`, specify the following fields:
+
[source,yaml]
----
output.elasticsearch:
 hosts: ["<your_elasticsearch_host>:9200"]
 protocol: "https"
 username: "metricbeat_writer"
 password: "<password>"
 ssl:
   certificate_authorities: ["elasticsearch-ca.pem"]

>    verification_mode: "certificate"

Reviewer 2: Why `verification_mode: "certificate"` here? Assuming they followed the steps up until here, the certificate ES is using has the correct hostname

Author: @jkakavas, is the verification mode unnecessary then?

----

   `hosts`:: Specifies the host where your Elasticsearch cluster is running.

   `protocol`:: Indicates the protocol to use when connecting to Elasticsearch.
   This value must be `https`.

   `username`:: Name of the user with privileges required to publish events to

>    Elasticsearch. The `metricbeat_writer` role that you created has these

Reviewer 2: ```suggestion
   Elasticsearch. The `metricbeat_writer` user that you created has these
```

Author: I'll change "role" to "user".

   privileges.

   `password`:: Password for the indicated `username`.

   `certificate_authorities`:: Indicates the path to your trusted CA.

. Configure the connection to Kibana.
+
Under `setup.kibana`, specify the following fields:
+
[source,yaml]
----
setup.kibana
 host: "https://<your_elasticsearch_host>:5601"
 ssl.enabled: true
 username: "metricbeat_setup"

>  password: "p@ssw0rd"

Reviewer: Unless kibana's server certificate is signed by a public CA or centrally managed CA, we probably need a CA setting here?

----

   `hosts`:: The URLs of the Elasticsearch instances to use for all your
   queries. Ensure that you include `https` in the URL.

   `username`:: Name of the user with privileges required to set up dashboards in Kibana. The `metricbeat_setup` role that you created has these privileges.

   `password`:: Password for the indicated `username`.

. Enable the `elasticsearch-xpack` module.
+
[source,shell]
----
./metricbeat modules enable elasticsearch-xpack
----

. Modify the `elasticsearch-xpack` module to use HTTPS.
+
Open `/modules.d/elasticsearch-xpack.yml` and specify the following fields:
+
[source,yaml]
----
- module: elasticsearch
 xpack.enabled: true
 period: 10s
 hosts: ["https://<your_elasticsearch_host>:9200"]
 username: "remote_monitoring_user"

>  password: "<password>"

Reviewer: I am not familiar with beats setup. So unless it shares configuration with the `output` section, I think we need the CA setting here as well.

----

   `hosts`:: Specifies the host where your Elasticsearch cluster is running.
   Ensure that you include `https` in the URL.

   `username`:: Name of the user with privileges to collect metric data. The

>    built-in `remote_monitoring_user` role has these privileges. Alternatively,

Reviewer 2: ```suggestion
   built-in `remote_monitoring_user` user has these privileges. Alternatively,
```

>    you can create a user and assign it the `remote_monitoring_user` role.

Reviewer 2: ```suggestion
   you can create a user and assign it the `monitoring_user` role.
```

The role name is `monitoring_user` https://www.elastic.co/guide/en/elasticsearch/reference/current/built-in-roles.html

Author: Great catch -- I'll change this to `monitoring_user` role.
